### PR TITLE
Always compile with comm thread and mini SeisSol enabled

### PR DESCRIPTION
**Pull request by davschneller**
As the title says, both the communication thread and mini SeisSol currently have compile-time variables to switch them on/off—but there's nothing from always supporting them. Thus, we do the following:

* introduce environment variables `SEISSOL_MINISEISSOL` and `SEISSOL_COMMTHREAD` (both boolean) which control if the two features are enabled, respectively
* also, introduce an additional parameter variable for mini SeisSol, in case we want to encode the determinism there.

### CODE REVIEW

File: src/Parallel/Helper.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+namespace seissol {
+template <typename T>
+void printCommThreadInfo(const T& mpiBasic) {
+  bool useThread = utils::Env::get<bool>("SEISSOL_COMMTHREAD", true);
+  if (mpiBasic.isSingleProcess()) {
+    logInfo(mpiBasic.rank()) << "Using polling for advancing MPI communication, due to having only "
+                                "one MPI rank running.";
+  }
+  if (useThread) {
+    logInfo(mpiBasic.rank()) << "Using a communication thread for advancing MPI communication.";
```
Comment (Thomas-Ulrich) [inline src/Parallel/Helper.hpp:14]: so if SEISSOL_COMMTHREAD=True and one node, we will get:
```
Using polling for advancing MPI communication, due to having only one MPI rank running.
Using a communication thread for advancing MPI communication.
```
Which is not what expected.

File: cmake/process_users_input.cmake [hunk 2]
```diff
@@ -6,11 +6,9 @@ set(GRAPH_PARTITIONING_LIB_OPTIONS parmetis parhip ptscotch)
 set_property(CACHE GRAPH_PARTITIONING_LIBS PROPERTY STRINGS ${GRAPH_PARTITIONING_LIB_OPTIONS})
 
 option(MPI "Use MPI parallelization" ON)
-option(MINI_SEISSOL "Use MiniSeisSol to compute node weights for load balancing" ON)
 option(OPENMP "Use OpenMP parallelization" ON)
 option(ASAGI "Use asagi for material input" OFF)
 option(MEMKIND "Use memkind library for hbw memory support" OFF)
-option(USE_IMPALA_JIT_LLVM "Use llvm version of impalajit" OFF)
```
Comment (Thomas-Ulrich) [inline cmake/process_users_input.cmake:13]: this is not really related to the PR?

File: src/SeisSol.cpp [hunk 3]
```diff
@@ -98,14 +99,19 @@ bool seissol::SeisSol::init(int argc, char* argv[]) {
...
+  if (seissol::useCommThread(MPI::mpi)) {
+    auto freeCpus = pinning.getFreeCPUsMask();
+    logInfo(rank) << "Communication thread affinity        :"
+                  << parallel::Pinning::maskToString(freeCpus);
+    if (parallel::Pinning::freeCPUsMaskEmpty(freeCpus)) {
+      logError()
+          << "There are no free CPUs left. Make sure to leave one for the communication thread. If "
+             "you want to run SeisSol without a communication thread (and instead use polling), "
+             "then try running with the environment variable \"SEISSOL_COMMTHREAD=false\". "
+             "However, for more balanced performance, we recommend using the communication thread.";
```
Comment (Thomas-Ulrich) [inline src/SeisSol.cpp:113]: what means more balanced performance? higher performance?

File: src/Initializer/InitProcedure/InitMesh.cpp [hunk 4]
```diff
@@ -58,24 +59,26 @@ static void readMeshPUML(const seissol::initializer::parameters::SeisSolParamete
...
-    nodeWeight = 1.0 / elapsedTime;
-
-    const auto summary = seissol::statistics::parallelSummary(nodeWeight);
-    logInfo(rank) << "Node weights: mean =" << summary.mean << " std =" << summary.std
-                  << " min =" << summary.min << " median =" << summary.median
-                  << " max =" << summary.max;
-
-    writer::MiniSeisSolWriter writer(seissolParams.output.prefix.c_str());
-    writer.write(elapsedTime, nodeWeight);
+  if (utils::Env::get<bool>("SEISSOL_MINISEISSOL", true) && seissolParams.mesh.miniseissol) {
```
Comment (Thomas-Ulrich) [inline src/Initializer/InitProcedure/InitMesh.cpp:62]: why to read the environment variable, if miniseissol is defined in the parameter file?
I would only keep the parameter file.
(by the way, commthread could also be defined in the parameter file).

**Review by Thomas-Ulrich — COMMENTED** [review]
I think it is a good idea. I ve noted a few things

**Comment by Thomas-Ulrich** [comment]
you also need to update the doc.
(e.g. supermucNG page, frontera, etc, parameters.par, do a search, there are a lot of occurences of COMMTHREAD)

File: src/Initializer/InitProcedure/InitMesh.cpp (hunk 4 above)
Comment (davschneller) [reply src/Initializer/InitProcedure/InitMesh.cpp:62]: Personally, I'd tend to keep them an environment variable at the moment, because they don't influence the simulation per se.
The mini SeisSol got into the parameter file as well, because it seemed to be needed in #427 IMO.

**Comment by davschneller** [comment]
> you also need to update the doc. (e.g. supermucNG page, frontera, etc, parameters.par, do a search, there are a lot of occurences of COMMTHREAD)

Yup, you're right. Probably, it's good if we enable versioning for the readthedocs before that; I'll look into it.

File: src/Parallel/Helper.hpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (davschneller) [reply src/Parallel/Helper.hpp]: Yup, I think so. I've replaced it.

File: cmake/process_users_input.cmake (hunk 2 above)
Comment (davschneller) [reply cmake/process_users_input.cmake:13]: Not really; it's only an unused option at the moment IIRC—or at least, since easi was removed as a submodule.

File: src/SeisSol.cpp (hunk 3 above)
Comment (davschneller) [reply src/SeisSol.cpp]: Removed that sentence.

File: src/Initializer/InitProcedure/InitMesh.cpp (hunk 4 above)
Comment (Thomas-Ulrich) [reply src/Initializer/InitProcedure/InitMesh.cpp:63]: I don't understand. I would only keep the environement variable, and remove seissolParams.mesh.miniseissol.
#427 can be fixed by using the environement variable, right?

Comment (davschneller) [reply src/Initializer/InitProcedure/InitMesh.cpp:63]: Yup, it can. Ok, I'll remove the parameter from the parameter file.

**Comment by davschneller** [comment]
Ok, the docs should be updated and the new environment variables documented.

File: Documentation/environment-variables.rst [hunk 6]
```diff
@@ -1,7 +1,25 @@
...
+By default, any SeisSol run with more than one MPI rank will use a communication thread to advance the MPI progress engine.
+For that, you will need to leave at least one thread vacant in your OpenMP thread placing map, cf. the SuperMUC-NG example below.
+
+If you do not want to use a communication thread, you may set `SEISSOL_COMMTHREAD=0`; then SeisSol polls on the progress from time to time.
+
+Load Balancing
+--------------
+
+When running with multiple ranks, SeisSol will estimate the performance of a node, to enable better load balancing for it.
+For that, it runs the so-called "Mini SeisSol" benchmark. As its name already hints at—it simulates a small test workload on each node;
```
Comment (Thomas-Ulrich) [inline Documentation/environment-variables.rst:18]: typo "at—it"

**Review by Thomas-Ulrich — APPROVED** [review]
LGTM, thanks!

File: Documentation/environment-variables.rst (hunk 6 above)
Comment (davschneller) [reply Documentation/environment-variables.rst]: Ok, replaced the "—" with a comma
